public/app/features/alerting/unified/components/rules/state-history/useRuleHistoryRecords.tsx

```diff
@@ -69,6 +71,46 @@ export function useRuleHistoryRecords(stateHistory?: DataFrameJSON, filter?: str
   }, [stateHistory, filter, theme]);
 }
 
+export function getRuleHistoryRecordsForPanel(stateHistory?: DataFrameJSON, filter?: string) {
+  if (!stateHistory) {
+    return { dataFrames: [] };
+  }
+  const theme = config.theme2;
+  // merge timestamp with "line"
+  const tsValues = stateHistory?.data?.values[0] ?? [];
+  const timestamps: number[] = isNumbers(tsValues) ? tsValues : [];
+  const lines = stateHistory?.data?.values[1] ?? [];
+
+  const logRecords = timestamps.reduce((acc: LogRecord[], timestamp: number, index: number) => {
+    const line = lines[index];
+    // values property can be undefined for some instance states (e.g. NoData)
+    if (isLine(line)) {
+      acc.push({ timestamp, line });
+    }
+
+    return acc;
+  }, []);
+
+  // group all records by alert instance (unique set of labels)
+  const logRecordsByInstance = groupBy(logRecords, (record: LogRecord) => {
+    return JSON.stringify(record.line.labels);
+  });
+
```

Comment:
This code looks similar to lines 27-44. Is it possible to extract a function for this? The code seems quite complex, and consolidating it into a single location would make it easier to maintain.